src/core/reference/include/openvino/reference/fake_convert.hpp

```diff
@@ -0,0 +1,329 @@
+    } else {
+        OPENVINO_THROW("Unsupported destination type.");
+    }
+    return true;
+}
+
+template <typename T, typename std::enable_if<std::is_same<T, float16>::value, bool>::type = true>
+void fake_convert(const T* data,
+                  const T* scale,
+                  const T* shift,
+                  T* out,
+                  const Shape& data_shape,
+                  const Shape& scale_shape,
+                  const Shape& shift_shape,
+                  const std::string& destination_type) {
+    const size_t element_count = shape_size(data_shape);
+    reference::fake_convert_details::apply_scale_shift<float16>(out,
+                                                                data,
+                                                                scale,
+                                                                shift,
+                                                                data_shape,
+                                                                scale_shape,
+                                                                shift_shape,
+                                                                false);
+    apply_conversion(out, out, element_count, destination_type);
+    reference::fake_convert_details::apply_scale_shift<float16>(out,
+                                                                out,
+                                                                scale,
+                                                                shift,
+                                                                data_shape,
+                                                                scale_shape,
+                                                                shift_shape,
+                                                                true);
+}
+
+template <typename T, typename std::enable_if<!std::is_same<T, float16>::value, bool>::type = true>
+void fake_convert(const T* data,
+                  const T* scale,
+                  const T* shift,
+                  T* out,
+                  const Shape& data_shape,
+                  const Shape& scale_shape,
+                  const Shape& shift_shape,
+                  const std::string& destination_type) {
+    const size_t element_count = shape_size(data_shape);
+    reference::fake_convert_details::apply_scale_shift<T>(out,
+                                                          data,
+                                                          scale,
+                                                          shift,
+                                                          data_shape,
+                                                          scale_shape,
+                                                          shift_shape,
+                                                          false);
+
+    std::vector<ov::float16> tmp_fp16;
+    tmp_fp16.reserve(element_count);
```

Comment:
Just allocate vector and then use it with size = 0, smells like UB. It would be better to use default initialization with size.

I thinks the allocation can be removed at all and apply conversion element by element in one loop instead make three.

Reply:
Is that confirmed that usage of reserve here is not correct? Updated to init with 0.f anyway.

Making all steps in a single loop will be considered as a follow up refactor.